### PR TITLE
chore(pg-meta): port functions tests

### PR DESCRIPTION
**Pull request by avallete**
## What kind of change does this PR introduce?

Continue the effort to migrate pg-meta into a self serving query lib

This PR:

1. Port over the tests for the "functions" query

### CODE REVIEW

File: packages/pg-meta/tsconfig.json [hunk 1]
```diff
@@ -1,7 +1,8 @@
 {
-  "extends": "tsconfig/base.json",
+  "extends": "../tsconfig/base.json",
   "compilerOptions": {
-    "module": "ESNext"
+    "module": "ESNext",
+    "moduleResolution": "bundler"
```
Comment (avallete) [inline packages/pg-meta/tsconfig.json:5]: **question**
Without this, I had types errors constantly showing up. Not sure if this is the best way to resolve this tough.

File: packages/pg-meta/test/db/utils.ts [hunk 2]
```diff
@@ -25,7 +25,8 @@ export async function createTestDatabase() {
       dbName,
       executeQuery: async <T = any>(query: string): Promise<T> => {
         try {
-          const result = (await testDb.unsafe(query)).map((v) => v)
+          const execResult = await testDb.unsafe(query)
+          const result = execResult.map((v) => v)
```
Comment (avallete) [inline packages/pg-meta/test/db/utils.ts:29]: Split it in two, making it easier to step in with the debugger if necessary

File: packages/pg-meta/test/db/00-init.sql [hunk 3]
```diff
@@ -1,4 +1,5 @@
-
+-- We apply our test seedings to template1 so every new created db will have the same structure
+\c template1
```
Comment (avallete) [inline packages/pg-meta/test/db/00-init.sql:2]: This is a little tricks to avoid the need to "seed" each new "created" database at jest/typescript level. That way each newly created database for each test will have our tables initialized.

File: packages/pg-meta/src/pg-meta-functions.ts [hunk 4]
```diff
@@ -212,8 +230,8 @@ function _generateCreateFunctionSql(
       config_params
         ? Object.entries(config_params)
             .map(
-              ([param, value]: string[]) =>
-                `SET ${param} ${value[0] === 'FROM CURRENT' ? 'FROM CURRENT' : 'TO ' + (value === '""' ? "''" : value)}`
+              ([param, value]) =>
+                `SET ${param} ${value === 'FROM CURRENT' ? 'FROM CURRENT' : 'TO ' + (value === '""' ? "''" : value)}`
```
Comment (avallete) [inline packages/pg-meta/src/pg-meta-functions.ts:234]: **question**

Had a type error here, either the `config_params` type is wrong but doesn't seems so. I've added a test to cover this case but might want another look.

File: packages/pg-meta/test/functions.test.ts [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+  expect(result1!.config_params).toEqual({
+    search_path: '""',
+    application_name: 'postgres.js',
+    work_mem: '8MB',
+  })
+
+  // Clean up
+  const { sql: removeSql1 } = await pgMeta.functions.remove(result1!)
+  await executeQuery(removeSql1)
+})
```
Comment (avallete) [inline packages/pg-meta/test/functions.test.ts:341]: New test to cover the FROM CURRENT case.

File: packages/pg-meta/src/pg-meta-functions.ts [hunk 6]
```diff
@@ -34,13 +58,11 @@ export const pgFunctionZod = z.object({
...
   security_definer: z.boolean(),
   config_params: z.union([z.record(z.string(), z.string()), z.null()]),
-})
+}) satisfies z.ZodType<PGFunction>
 
 export const pgFunctionArrayZod = z.array(pgFunctionZod)
 export const pgFunctionOptionalZod = z.optional(pgFunctionZod)
 
-export type PGFunction = z.infer<typeof pgFunctionZod>
-
```
Comment (soedirgo) [inline packages/pg-meta/src/pg-meta-functions.ts:43]: Pourquoi ?

File: packages/pg-meta/tsconfig.json (hunk 1 above)
Comment (soedirgo) [reply packages/pg-meta/tsconfig.json:5]: Seems fine, let's go with it

File: packages/pg-meta/src/pg-meta-functions.ts (hunk 6 above)
Comment (avallete) [reply packages/pg-meta/src/pg-meta-functions.ts:43]: Good catch, that's a remaining artifact of me trying out some things to debug the types I had in my tests 😅 

Reverted this back to being infered by zod schema.

File: packages/pg-meta/src/sql/functions.ts [hunk 7]
```diff
@@ -80,15 +80,16 @@ from
...
+        'has_default', COALESCE(has_default, false)
       )) as args
     from
       (
         select
           oid,
           unnest(arg_modes) as mode,
           unnest(arg_names) as name,
-          unnest(arg_types) as type_id,
+          unnest(arg_types)::int8 as type_id,
```
Comment (soedirgo) [inline packages/pg-meta/src/sql/functions.ts:92]: Let's keep this as `oid` instead of `int8`

File: packages/pg-meta/src/sql/functions.ts [hunk 8]
```diff
@@ -80,15 +80,16 @@ from
         'mode', t2.mode,
         'name', name,
         'type_id', type_id,
-        'has_default', has_default
+        -- Cast null into false boolean
+        'has_default', COALESCE(has_default, false)
```
Comment (soedirgo) [inline packages/pg-meta/src/sql/functions.ts:84]: Is there ever a case where `has_default` is `null`?

Comment (soedirgo) [reply packages/pg-meta/src/sql/functions.ts:84]: Would've expected there to be none according to this https://github.com/supabase/supabase/blob/d78768cec464f420b765290208fd60ed98b1085d/packages/pg-meta/src/sql/functions.ts#L20

File: packages/pg-meta/src/pg-meta-functions.ts (hunk 4 above)
Comment (soedirgo) [reply packages/pg-meta/src/pg-meta-functions.ts:212]: Yeah I think the `value[0]` was wrong before

File: packages/pg-meta/src/sql/functions.ts (hunk 7 above)
Comment (avallete) [reply packages/pg-meta/src/sql/functions.ts:92]: Can do but will have to convert the number in zod type to a string, because oid get out as a string with `postgres.`

File: packages/pg-meta/src/sql/functions.ts (hunk 8 above)
Comment (avallete) [reply packages/pg-meta/src/sql/functions.ts:84]: Yup some of them get null values without that.

Screenshot 2025-02-11 at 19 32 06

**Comment by avallete** [comment]
Updated things out to use `pg` instead of `postgres` since that's what's will be executed on the `postgres-meta` side.

That come from a discussion oustide of github on slack, here is a summary for context:

Refactor to use of the `pg` client for testing while preparing for potential future client changes. The main reasons are:  

1. **Consistency and Maintainability**: Using `pg` minimizes diffs and aligns with query execution in `postgres-meta` with the `executeQuery` endpoint
2. **Type Handling Improvements**: Discrepancies arose from custom type parsing (e.g., `INT8` as `number` or `string`), leading to inconsistent test results. To address this, `pg` is configured with type overrides similar to those in `postgres-meta`, ensuring consistent behavior.  
3. **Zod Integration**: Zod is kept to validate runtime values, catching edge cases like `null` defaults and `bigint > safeInteger`. This allows safe future client switches by flagging type differences during tests.  
4. **Testing Strategy**: Tests continue to use `toMatchInlineSnapshot` since we’re only using the `pg` client for now. However, the refactor prepares the codebase to easily switch clients later and catch any differences in runtime behavior.  

This approach keeps the current tests reliable while ensuring flexibility for future changes.

**Review by soedirgo — APPROVED** [review]
